### DISCUSSION TITLE
Creating a DatePicker with a year or decade view

### DISCUSSION THREAD

**Discussion by mattwendzina** [adobe/react-spectrum#3950] [Q&A]
Hi,
I've had a great experience creating a DatePicker component and am grateful for the fantastic documentation.

However, I've hit a bit of a blocker in trying to create a Datepicker with both a yearly view and a decade view. That is, where I can allow users to select a month or a year rather than a day. And where the Previous and Next buttons will enable me to go forwards/back a year/decade respectively.

Does anyone know how this can be achieved and can point me to any examples?

Many thanks in advance

**Comment by LFDanLu** [comment]
Just to clarify, are you trying to make a DatePicker with a Calendar grid made up of months/years like this?
<img width="204" alt="image" src="https://user-images.githubusercontent.com/9661127/215914528-8adb79a2-b31e-48c8-bd78-4915423742a6.png">

Or are you trying to extend a typical month Calendar grid with additional controls for month/year selection?

**Reply by mattwendzina** [reply]
Hi @LFDanLu - Yep, like that image you've shown. So that the up/down arrows (in that example) would toggle the year. And then one step up from that, a decade view that allows you to select a year and the arrows toggle each decade/set of years.

**Reply by LFDanLu** [reply]
I see, my initial gut reaction is that this isn't possible out of the box with the hooks as is. As you can see [here](https://github.com/adobe/react-spectrum/blob/c9dcf67a5f31047819d5c2ef5d67b8c8ae0a0f11/packages/%40react-aria/calendar/src/useCalendarGrid.ts#L85-L112), `useCalendarGrid` updates the focus by an increment of a day when using the arrow keys (at least when using ArrowLeft/ArrowRight) by using [these callbacks](https://github.com/adobe/react-spectrum/blob/c9dcf67a5f31047819d5c2ef5d67b8c8ae0a0f11/packages/%40react-stately/calendar/src/useCalendarState.ts#L190-L195) returned by `useCalendarState`. I imagine that would already be a blocker since left/right in the example grid above should move by month increments.

**Reply by mattwendzina** [reply]
That's a shame. Are these hooks that Adobe would consider adding at a future point?

**Reply by devongovett** [reply]
Keep in mind that you could build this yourself and reuse the calendar state. This example shows a custom month/year dropdown that does this. Your example has a bit more complex UI, but it should be possible in a similar way. https://codesandbox.io/s/affectionate-rosalind-tdm323?file=/src/Calendar.js

**Reply by mattwendzina** [reply]
Thanks for this @devongovett, really helpful! I've managed to get a yearly view working now, with the DatePicker 🙌  
https://codesandbox.io/s/xenodochial-curie-9k38nz?file=/src/DatePicker.js

**Reply by PixeledCode** [reply]
Were you able to handle focus change using navigation keys?

**Reply by TylerLHenderson** [reply]
Hey @mattwendzina, your codesandbox link is gone. 
Were you able to get the year grid view working or did you go with the year dropdown select version?

**Reply by ChernYeeLow** [reply]
I've created a customized calendar with month/year view. But when I integrate the calendar to RAC date picker component, the value of the DatePicker fails to pass to the Custom Calendar component. When the user selects a date in the Calendar, the popover also does not close, and the value of the calendar does not pass back to the DatePicker component. Any ideas?

**Reply by keving2301** [reply]
I am running into this same issue. I was able to get a little farther in the implementation where I was able to implement a month/year picker component and use the setFocusedDate to make the Grid focus on the correct month/year after the selection. However new challenges come with a dual calendar implementation. Does anyone have a hint on how (if possible) to manipulate the visibleRange dates so that I am able to assign month and year from a left or right calendar?

**Comment by cssinate** [comment]
I just wanted to add an additional voice of support for expanding the functionality of this calendar. I work for a company that builds software for retirement communities. Selecting a birth date from a calendar is something our users like to be able to do. If you look at the example here: https://projects.wojtekmaj.pl/react-calendar/

Clicking the month name at the top of the calendar, brings you to a month selector. Clicking the year at the top of the month selector brings you to a year selector. In the year selector, clicking the back and forth buttons runs the user through decades. Having built-in functionality like this would be really helpful.

**Reply by devongovett** [reply]
For entering birth dates, have you looked at using a [DateField](https://react-spectrum.adobe.com/react-aria/DateField.html) instead of a Calendar? IMO it's much easier to enter dates you already know with a keyboard rather than clicking 17 times to get to the right part of a calendar. Calendar is much more useful for a browsing experience when you aren't sure what date you'll select. You can also combine the two with a [DatePicker](https://react-spectrum.adobe.com/react-aria/DatePicker.html).

**Reply by cssinate** [reply]
Yes, we're doing both a text input and a calendar. Our user research shows that a not-insignificant portion of our users prefer to use the calendar rather than the input field when they're on iPad.
